sympy/holonomic/linearsolver.py

```diff
@@ -6,6 +6,7 @@
 from sympy.matrices.common import ShapeError
 from sympy.matrices.dense import MutableDenseMatrix
 
+from ..matrices.reductions import _rref
```

Comment:
Is a relative import safe here? I would opt for an absolute import.

Reply:
Needs to be relative since `_rref` is private (not in `matrices/__init__.py` `__all__`), which is a valid question, cross-submodule private imports?

Reply:
Yes...if private functions are cross-imported that defeats the point of having them private. This is another reason to prefer calling `...Matrix._eval_rref(...)` directly on the class instead of hoisting everything to non-exported functions.

Reply:
The main thing I don't like about having for example `rref` and `_eval_rref` is that you now need two methods for a single operation. I understand the reasoning behind it but I don't think in this case it is warranted since if you are going to be overloading `rref` you will probably do it at the top one anyway. 

As for avoiding private import from another submodule the other option is to make a public rref function exported from matrices which does not simpify (which is what causes these matrices to raise).

Reply:
Another reason for the `_eval_*` methods, was that if you subclass `rref` directly, the docstring goes away, which means you have ugly code to copy it. In addition, you need to reimpliment any error/bounds checking and argument sanitization. It may not be a big deal on `rref` specifically, but I think consistency is a win overall.

Reply:
True, but on the other hand then you have to duplicate the function signature and any default assignments and maintain it in two places. I've solved this particular case by absolute import which I was just being an idiot for not seeing it but I think the discussion for the presence of `_eval_*` methods and separate standalone non-method functions is something you guys need to hash out among the elders. I personally do see a use for individual standalone functions regardless of how the class methods eventually get to them so would like to continue pulling them out but it sounds like the discussion for the next matrix API should really get going...